<a href="https://colab.research.google.com/github/arvndms/ICT-DSA_ACTIVITY/blob/main/Home_Loan_Predicition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Libraries

In [856]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

In [857]:
#loading dataset
train = pd.read_csv('/content/home_loan_train.csv')
test = pd.read_csv('/content/home_loan_test.csv')

Working with train data

In [858]:
train.head()

Loan_ID Gender Married  ... Credit_History Property_Area Loan_Status
0  LP001002   Male      No  ...            1.0         Urban           Y
1  LP001003   Male     Yes  ...            1.0         Rural           N
2  LP001005   Male     Yes  ...            1.0         Urban           Y
3  LP001006   Male     Yes  ...            1.0         Urban           Y
4  LP001008   Male      No  ...            1.0         Urban           Y

[5 rows x 13 columns]

In [859]:
train.dtypes

Loan_ID               object
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

In [860]:
train.describe()

ApplicantIncome  CoapplicantIncome  ...  Loan_Amount_Term  Credit_History
count       614.000000         614.000000  ...         600.00000      564.000000
mean       5403.459283        1621.245798  ...         342.00000        0.842199
std        6109.041673        2926.248369  ...          65.12041        0.364878
min         150.000000           0.000000  ...          12.00000        0.000000
25%        2877.500000           0.000000  ...         360.00000        1.000000
50%        3812.500000        1188.500000  ...         360.00000        1.000000
75%        5795.000000        2297.250000  ...         360.00000        1.000000
max       81000.000000       41667.000000  ...         480.00000        1.000000

[8 rows x 5 columns]

In [861]:
train.isna().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [862]:
test.head()

Loan_ID Gender Married  ... Loan_Amount_Term Credit_History Property_Area
0  LP001015   Male     Yes  ...            360.0            1.0         Urban
1  LP001022   Male     Yes  ...            360.0            1.0         Urban
2  LP001031   Male     Yes  ...            360.0            1.0         Urban
3  LP001035   Male     Yes  ...            360.0            NaN         Urban
4  LP001051   Male      No  ...            360.0            1.0         Urban

[5 rows x 12 columns]

In [863]:
test.isna().sum()

Loan_ID               0
Gender               11
Married               0
Dependents           10
Education             0
Self_Employed        23
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount            5
Loan_Amount_Term      6
Credit_History       29
Property_Area         0
dtype: int64

In [864]:
#spliting the target feature
X = train.drop('Loan_Status',axis=1)
y = train['Loan_Status']

In [865]:
#filling Dependents with mode
X['Dependents'] = X['Dependents'].fillna(X['Dependents'].mode()[0])
test['Dependents'] = test['Dependents'].fillna(test['Dependents'].mode()[0])

In [866]:
#filling LoanAmount with median
X['LoanAmount'] = X['LoanAmount'].fillna(X['LoanAmount'].median())
test['LoanAmount'] = test['LoanAmount'].fillna(test['LoanAmount'].median())

In [867]:
X['Loan_Amount_Term'] = X['Loan_Amount_Term'].fillna(X['Loan_Amount_Term'].mode()[0])
test['Loan_Amount_Term'] = test['Loan_Amount_Term'].fillna(test['Loan_Amount_Term'].mode()[0])

In [868]:
X['Credit_History'] = X['Credit_History'].fillna(X['Credit_History'].mode()[0])
test['Credit_History'] = test['Credit_History'].fillna(test['Credit_History'].mode()[0])

In [869]:
X['Gender'] = X['Gender'].fillna(X['Gender'].mode()[0])
X['Married'] = X['Married'].fillna(X['Married'].mode()[0])

In [870]:
test['Gender'] = test['Gender'].fillna(test['Gender'].mode()[0])

In [871]:
X['Self_Employed'] = X['Self_Employed'].fillna(X['Self_Employed'].mode()[0])
test['Self_Employed'] = test['Self_Employed'].fillna(test['Self_Employed'].mode()[0])

In [872]:
X['Dependents'] = X['Dependents'].replace('3+',3)
test['Dependents'] = test['Dependents'].replace('3+',3)

In [873]:
X.isna().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
dtype: int64

Encoding

In [874]:
le = LabelEncoder()
cols_to_encode = ['Gender','Married','Education','Self_Employed','Property_Area'] #cols that need to be encoded

In [875]:
for cols in cols_to_encode:
  X[cols] = le.fit_transform(X[cols])
  test[cols] = le.fit_transform(test[cols])

In [876]:
X.head()

Loan_ID  Gender  Married  ... Loan_Amount_Term  Credit_History  Property_Area
0  LP001002       1        0  ...            360.0             1.0              2
1  LP001003       1        1  ...            360.0             1.0              0
2  LP001005       1        1  ...            360.0             1.0              2
3  LP001006       1        1  ...            360.0             1.0              2
4  LP001008       1        0  ...            360.0             1.0              2

[5 rows x 12 columns]

In [877]:
y = y.map({'Y':1,'N':0})

In [878]:
X = X.drop('Loan_ID',axis=1)

In [879]:
test =  test.drop('Loan_ID',axis=1)

In [880]:
X.head()

Gender  Married Dependents  ...  Loan_Amount_Term  Credit_History  Property_Area
0       1        0          0  ...             360.0             1.0              2
1       1        1          1  ...             360.0             1.0              0
2       1        1          0  ...             360.0             1.0              2
3       1        1          0  ...             360.0             1.0              2
4       1        0          0  ...             360.0             1.0              2

[5 rows x 11 columns]

In [881]:
test.isna().sum()

Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
dtype: int64

In [882]:
test.head()

Gender  Married Dependents  ...  Loan_Amount_Term  Credit_History  Property_Area
0       1        1          0  ...             360.0             1.0              2
1       1        1          1  ...             360.0             1.0              2
2       1        1          2  ...             360.0             1.0              2
3       1        1          2  ...             360.0             1.0              2
4       1        0          0  ...             360.0             1.0              2

[5 rows x 11 columns]

Outliers

outliers in train data

In [883]:
numeric_cols = ['ApplicantIncome','CoapplicantIncome','LoanAmount','Loan_Amount_Term','Credit_History']
for col in numeric_cols:
  Q1 = X[col].quantile(0.25)
  Q3 = X[col].quantile(0.75)
  IQR = Q3 - Q1
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR

  outliers = X[(X[col] < lower_bound) | (X[col] > upper_bound)]
  print(f"Outliers in {col}:")
  print(f"No of Outliers:{outliers.shape[0]}")



Outliers in ApplicantIncome:
No of Outliers:50
Outliers in CoapplicantIncome:
No of Outliers:18
Outliers in LoanAmount:
No of Outliers:41
Outliers in Loan_Amount_Term:
No of Outliers:88
Outliers in Credit_History:
No of Outliers:89


In [884]:
#capping outliers
for col in numeric_cols:
  Q1 = X[col].quantile(0.25)
  Q3 = X[col].quantile(0.75)
  IQR = Q3 - Q1
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR

  X[col] = X[col].clip(lower_bound,upper_bound)

In [885]:
# again checking for outliers
for col in numeric_cols:
  Q1 = X[col].quantile(0.25)
  Q3 = X[col].quantile(0.75)
  IQR = Q3 - Q1
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR

  outliers = X[(X[col] < lower_bound) | (X[col] > upper_bound)]
  print(f"Outliers in {col}:")
  print(f"No of Outliers:{outliers.shape[0]}")


Outliers in ApplicantIncome:
No of Outliers:0
Outliers in CoapplicantIncome:
No of Outliers:0
Outliers in LoanAmount:
No of Outliers:0
Outliers in Loan_Amount_Term:
No of Outliers:0
Outliers in Credit_History:
No of Outliers:0


outliers in test data

In [886]:
numeric_cols = ['ApplicantIncome','CoapplicantIncome','LoanAmount','Loan_Amount_Term','Credit_History']
for col in numeric_cols:
  Q1 = test[col].quantile(0.25)
  Q3 = test[col].quantile(0.75)
  IQR = Q3 - Q1
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR

  outliers = test[(test[col] < lower_bound) | (test[col] > upper_bound)]
  print(f"Outliers in {col}:")
  print(f"No of Outliers:{outliers.shape[0]}")

Outliers in ApplicantIncome:
No of Outliers:32
Outliers in CoapplicantIncome:
No of Outliers:8
Outliers in LoanAmount:
No of Outliers:18
Outliers in Loan_Amount_Term:
No of Outliers:50
Outliers in Credit_History:
No of Outliers:59


In [887]:
for col in numeric_cols:
  Q1 = X[col].quantile(0.25)
  Q3 = X[col].quantile(0.75)
  IQR = Q3 - Q1
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR

  test[col] = test[col].clip(lower_bound,upper_bound)

Spliting train data train_test_split

In [888]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

Logistic Regression and Random forest Classifier

In [889]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score,confusion_matrix,f1_score,recall_score,precision_score

In [890]:
lr = LogisticRegression(max_iter=5000)
lr.fit(X_train,y_train)
lr_pred = lr.predict(X_test)

In [891]:
print('Accuracy Score',accuracy_score(y_test,lr_pred))
print('F1 Score',f1_score(y_test,lr_pred))
print('Recall Score',recall_score(y_test,lr_pred))
print('Precision Score',precision_score(y_test,lr_pred))
print('Confusion Matrix',confusion_matrix(y_test,lr_pred))

Accuracy Score 0.6585365853658537
F1 Score 0.7920792079207921
Recall Score 1.0
Precision Score 0.6557377049180327
Confusion Matrix [[ 1 42]
 [ 0 80]]


In [892]:
rf = RandomForestClassifier(n_estimators=500,max_depth=8,random_state=42)
rf.fit(X_train,y_train)
rf_pred = rf.predict(X_test)

In [893]:
print('Accuracy Score',accuracy_score(y_test,rf_pred))
print('F1 Score',f1_score(y_test,rf_pred))
print('Recall Score',recall_score(y_test,rf_pred))
print('Precision Score',precision_score(y_test,rf_pred))
print('Confusion Matrix',confusion_matrix(y_test,rf_pred))

Accuracy Score 0.6504065040650406
F1 Score 0.7860696517412935
Recall Score 0.9875
Precision Score 0.6528925619834711
Confusion Matrix [[ 1 42]
 [ 1 79]]


In [894]:
test.head()

Gender  Married Dependents  ...  Loan_Amount_Term  Credit_History  Property_Area
0       1        1          0  ...             360.0             1.0              2
1       1        1          1  ...             360.0             1.0              2
2       1        1          2  ...             360.0             1.0              2
3       1        1          2  ...             360.0             1.0              2
4       1        0          0  ...             360.0             1.0              2

[5 rows x 11 columns]

In [895]:
y_test_pred = rf.predict(test)

In [896]:
y_test_pred = pd.Series(y_test_pred).map({1:'Y',0:'N'})

In [897]:
sample = pd.read_csv('/content/sample_submission_49d68Cx.csv')
sample.head()

Loan_ID Loan_Status
0  LP001015           N
1  LP001022           N
2  LP001031           N
3  LP001035           N
4  LP001051           N

In [898]:
submit = pd.DataFrame({
    'Loan_ID':sample['Loan_ID'],
    'Loan_Status':y_test_pred
})
submit.to_csv('submit.csv',index=False)

In [899]:
submission = pd.read_csv('/content/submit.csv')
submission.head()

Loan_ID Loan_Status
0  LP001015           Y
1  LP001022           Y
2  LP001031           Y
3  LP001035           Y
4  LP001051           Y

In [900]:
y_test_pred_lr = lr.predict(test)

In [901]:
y_test_pred_lr = pd.Series(y_test_pred).map({1:'Y',0:'N'})

In [902]:
submit_lr = pd.DataFrame({
    'Loan_ID':sample['Loan_ID'],
    'Loan_Status':y_test_pred
})
submit_lr.to_csv('submit_lr.csv',index=False)

In [903]:
submission_lr = pd.read_csv('/content/submit_lr.csv')
submission_lr.head()

Loan_ID Loan_Status
0  LP001015           Y
1  LP001022           Y
2  LP001031           Y
3  LP001035           Y
4  LP001051           Y

Both of the models are showing similar result.
But i am going with Logistic Regressor cause of it slightly better perfomance
- Accuracy Score 0.6585365853658537
- F1 Score 0.7920792079207921
- Recall Score 1.0
- Precision Score 0.6557377049180327
- Confusion Matrix [[ 1 42]
 [ 0 80]]